_posts/2025-05-05-plan-execute-agent.md

```diff
@@ -0,0 +1,447 @@
+
+## What's new in OpenSearch 3.0's agent framework?
+
+OpenSearch 3.0 introduces an improved agent framework that enhances how you can interact with your data and address complex problems. The primary addition is a plan-execute-reflect agent designed to handle multi-step tasks methodically.
+
+The new agent framework includes several notable improvements:
+
+- A plan-execute-reflect agent that systematically breaks down and executes complex multi-step tasks
+- Native MCP (Model Context Protocol) support for integration with external AI tools
+- Enhanced tool selection and execution capabilities that select appropriate tools for each task
+- Asynchronous execution support for managing long-running tasks efficiently
+
+In this blog, we'll dive into a practical observability use case that showcases the true power of this new agent framework. You'll see firsthand how it can automatically investigate service failures in a microservices application—a typically time-consuming and complex process. You'll see how the agent can:
+
+* Break down a complex investigation into logical steps
+* Automatically select and use appropriate tools
+* Analyze multiple data sources (logs, traces, metrics)
+* Provide clear, actionable insights
+
+Whether you're managing cloud infrastructure, developing applications, or building AI-powered solutions, this new capability can help streamline your workflows and enhance your problem-solving capabilities.
+
+Now let's explore the capabilities and workflow of the plan-execute-reflect agent.
+
+## What is the plan-execute-reflect agent?
+
+The [plan-execute-reflect](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/agents/plan-execute-reflect/) agent is a long-running agent designed for complex, multi-step tasks. This agent is capable of breaking down a complex task into a series of simple steps (plan), executing each step, and optimizing its plan based on intermediate step results. It uses a separate executor agent for the execution of substeps.
+
+Key features of this agent include the following:
+
+* Adaptive planning that evolves based on intermediate results
+* Flexibility to use different models for planning and execution phases by specifying your own executor agent during registration (uses a [conversational](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/agents/conversational/) agent by default)
+* Asynchronous execution capability for handling long-running workflows in the background (when `async=true`)
+* Built-in MCP client functionality allowing connections to multiple MCP servers
+* Standardized tool communication through function calling
+* Support for custom prompts, enabling specialization for specific use cases such as Observability Agent or Research Agent (using WebSearchTool)
```

Comment:
"such as building an observability agent or research agent (using the WebSearchTool)"?